mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/search/CriterionFactory.java

```diff
@@ -0,0 +1,115 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.vault.search;
+
+import static org.apache.james.vault.search.DeletedMessageField.DELETION_DATE;
+import static org.apache.james.vault.search.DeletedMessageField.DELIVERY_DATE;
+import static org.apache.james.vault.search.DeletedMessageField.HAS_ATTACHMENT;
+import static org.apache.james.vault.search.DeletedMessageField.ORIGIN_MAILBOXES;
+import static org.apache.james.vault.search.DeletedMessageField.RECIPIENTS;
+import static org.apache.james.vault.search.DeletedMessageField.SENDER;
+import static org.apache.james.vault.search.DeletedMessageField.SUBJECT;
+
+import java.time.ZonedDateTime;
+import java.util.List;
+
+import org.apache.james.core.MailAddress;
+import org.apache.james.mailbox.model.MailboxId;
+
+public interface CriterionFactory {
+
+    class StringCriterionFactory {
+
+        private final DeletedMessageField<String> deletedMessageField;
+
+        private StringCriterionFactory(DeletedMessageField<String> deletedMessageField) {
+            this.deletedMessageField = deletedMessageField;
+        }
+
+        public Criterion<String> contains(String subString) {
+            return new Criterion<>(deletedMessageField, ValueMatcher.SingleValueMatcher.contains(subString));
```

Comment:
The fact that you reference ValueMatcher into CriterionFactory and CriterionFactory in ValueMatcher should trigger a warning: the separation of concern is probably not good.
BTW, I have hard time figuring out the design you implemented, it should be possible to simplify the design again.